stock_available_to_promise_release/models/stock_move.py

```diff
@@ -419,3 +487,82 @@ def _assign_picking_post_process(self, new=False):
         if priorities:
             self.picking_id.write({"priority": max(priorities)})
         return res
+
+    def _get_chained_moves_iterator(self, chain_field):
+        """Return an iterator on the moves of the chain.
+
+        The iterator returns the moves in the order of the chain.
+        The loop into the iterator is the current moves.
+        """
+        moves = self
+        while moves:
+            yield moves
+            moves = moves.mapped(chain_field)
+
+    def unrelease(self, safe_unrelease=False):
+        """Unrelease unreleasavbe moves
+
+        If safe_unrelease is True, the unreleasaable moves for which the
+        processing has already started will be ignored
+        """
+        moves_to_unrelease = self.filtered(lambda m: m._is_unreleaseable())
+        if safe_unrelease:
+            moves_to_unrelease = self.filtered("unrelease_allowed")
+        moves_to_unrelease._check_unrelease_allowed()
+        moves_to_unrelease.write({"need_release": True})
+        impacted_picking_ids = set()
+        for move in moves_to_unrelease:
+            iterator = move._get_chained_moves_iterator("move_orig_ids")
+            next(iterator)  # skip the current move
+            for origin_moves in iterator:
+                origin_moves = origin_moves.filtered(
+                    lambda m: m.state not in ("done", "cancel")
+                )
+                if origin_moves:
+                    origin_moves = move._split_origins(origin_moves)
+                    impacted_picking_ids.update(origin_moves.mapped("picking_id").ids)
+                    # avoid to propagate cancel to the original move
+                    origin_moves.write({"propagate_cancel": False})
+                    origin_moves._action_cancel()
```

Comment:
You have to cancel starting of the first origin of the chain and go down the chain. Easier it to make a complete recordset and cancel them all together.
So in case of pick/pack/ship, you cannot cancel pack then pick. You have to cancel pick, then pack, or pick+pack together.